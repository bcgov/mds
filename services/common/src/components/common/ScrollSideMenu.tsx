import React, { FC, useEffect } from "react";
import { useParams, useHistory, useLocation } from "react-router-dom";
import { Anchor } from "antd";
import { ISideMenuOption } from "../../interfaces/common/sideMenuOption.interface";

/**
 * @constant ScrollSideMenu renders react children with an active indicator if the id is in the url.
 */

interface ScrollSideMenuProps {
  menuOptions: ISideMenuOption[];
  featureUrlRoute: any;
  featureUrlRouteArguments: (string | number)[];
  tabSection?: string;
}

export const ScrollSideMenu: FC<ScrollSideMenuProps> = ({ tabSection = "", ...props }) => {
  const history = useHistory();
  const location = useLocation();
  const { tab } = useParams<{
    tab: string;
  }>();
  let urlRoute = undefined;

  const updateUrlRoute = (route) => {
    urlRoute = props.featureUrlRoute(...props.featureUrlRouteArguments, route);
    if (route === history.location.hash) {
      return;
    }

    history.push(urlRoute, { currentActiveLink: route });
  };

  useEffect(() => {
    // If the user loads the page with a hash in the URL, start them off at the corresponding feature section.
    // Note: Because Keycloak authorization adds params to the URL when it redirects, props.location.hash
    // will be contaminated with extra params that we don't need. All we want is the hash that corresponds
    // to the feature section, so we must parse it out. If the hash is "#state", we must ignore it (see example).
    // For example: #blasting&state=bd74ea1c-09e5-4d7e-810f-d3558969293a&session_state=1c577088-15a8-4ae2-...
    let link = !location.hash.startsWith("#state") ? location.hash : undefined;
    if (!link) {
      return;
    }

    // Extracts "#blasting" from "#blasting&state=bd74ea1c-09e5-4d7e-810f-d...", for example.
    if (link.includes("&")) {
      link = link.substring(0, link.indexOf("&"));
    }

    updateUrlRoute(link);
    document.querySelector(link)?.scrollIntoView();
  }, []);

  const handleAnchorOnClick = (e, link) => {
    e.preventDefault();
    updateUrlRoute(link.href);
  };

  const handleAnchorOnChange = (currentActiveLink) => {
    if (
      (history.action === "POP" && currentActiveLink === history.location.hash) ||
      tab !== tabSection
    ) {
      return;
    }

    history.replace(urlRoute, { currentActiveLink });
  };

  return (
    <div>
      <Anchor
        affix={false}
        offsetTop={180}
        onChange={handleAnchorOnChange}
        onClick={handleAnchorOnClick}
      >
        {props.menuOptions.map(({ href, title, icon }) => {
          const titleElement = (
            <div className="ellipsis-text">
              {icon && <span className="margin-medium--right">{icon}</span>}
              {title}
            </div>
          );
          return (
            <Anchor.Link
              key={title}
              href={`#${href}`}
              title={titleElement}
              className="now-menu-link"
            />
          );
        })}
      </Anchor>
    </div>
  );
};

export default ScrollSideMenu;
