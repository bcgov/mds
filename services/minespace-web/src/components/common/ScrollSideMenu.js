import React, { Component } from "react";
import { withRouter } from "react-router-dom";
import { PropTypes } from "prop-types";
import { Anchor } from "antd";

/**
 * @constant ScrollSideMenu renders react children with an active indicator if the id is in the url.
 */

const propTypes = {
  menuOptions: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.string)).isRequired,
  featureUrlRoute: PropTypes.func.isRequired,
  featureUrlRouteArguments: PropTypes.arrayOf(PropTypes.oneOf([PropTypes.string, PropTypes.number]))
    .isRequired,
  history: PropTypes.shape({
    push: PropTypes.func,
    replace: PropTypes.func,
    action: PropTypes.string,
    location: PropTypes.shape({ hash: PropTypes.string }),
  }).isRequired,
  location: PropTypes.shape({ hash: PropTypes.string }).isRequired,
  match: PropTypes.shape({
    params: {
      id: PropTypes.string,
    },
  }).isRequired,
  tabSection: PropTypes.string,
};

const defaultProps = { tabSection: "" };

export class ScrollSideMenu extends Component {
  state = {};

  // eslint-disable-next-line react/sort-comp
  static urlRoute = undefined;

  componentDidMount() {
    // If the user loads the page with a hash in the URL, start them off at the corresponding feature section.
    // Note: Because Keycloak authorization adds params to the URL when it redirects, props.location.hash
    // will be contaminated with extra params that we don't need. All we want is the hash that corresponds
    // to the feature section, so we must parse it out. If the hash is "#state", we must ignore it (see example).
    // For example: #blasting&state=bd74ea1c-09e5-4d7e-810f-d3558969293a&session_state=1c577088-15a8-4ae2-...
    let link =
      this.props.location &&
        this.props.location.hash &&
        !this.props.location.hash.startsWith("#state")
        ? this.props.location.hash
        : undefined;

    if (!link) {
      return;
    }

    // Extracts "#blasting" from "#blasting&state=bd74ea1c-09e5-4d7e-810f-d...", for example.
    if (link.includes('&')) {
      link = link.substr(0, link.indexOf("&"));
    }

    this.updateUrlRoute(link);

    document.querySelector(link)?.scrollIntoView()
  }

  handleAnchorOnClick = (e, link) => {
    e.preventDefault();
    this.updateUrlRoute(link.href);
  };

  handleAnchorOnChange = (currentActiveLink) => {
    if (
      (this.props.history.action === "POP" &&
        currentActiveLink === this.props.history.location.hash) ||
      this.props.match.params.tab !== this.props.tabSection
    ) {
      return;
    }

    this.props.history.replace(this.urlRoute, { currentActiveLink });
  };

  updateUrlRoute = (route) => {
    this.urlRoute = this.props.featureUrlRoute(...this.props.featureUrlRouteArguments, route);

    if (route === this.props.history.location.hash) {
      return;
    }

    this.props.history.push(this.urlRoute, { currentActiveLink: route });
  };

  render() {
    return (
      <div>
        <Anchor
          affix={false}
          offsetTop={160}
          onChange={this.handleAnchorOnChange}
          onClick={this.handleAnchorOnClick}
        >
          {this.props.menuOptions.map(({ href, title }) => (
            <Anchor.Link key={title} href={`#${href}`} title={title} className="now-menu-link" />
          ))}
        </Anchor>
      </div>
    );
  }
}

ScrollSideMenu.propTypes = propTypes;
ScrollSideMenu.defaultProps = defaultProps;

export default withRouter(ScrollSideMenu);
