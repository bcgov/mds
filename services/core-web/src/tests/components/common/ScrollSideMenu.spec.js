import React from "react";
import { shallow } from "enzyme";
import ScrollSideMenu from "@/components/common/ScrollSideMenu";

const props = {};

const setupProps = () => {
  props.menuOptions = [{ href: "section-1", title: "Section #1" }];
  props.featureUrlRoute = jest.fn();
  props.featureUrlRouteArguments = ["1stGuid", "2ndGuid"];
  props.match = { params: { id: "1234-4567-xwqy-j765" } };
  props.history = {
    push: jest.fn(),
    replace: jest.fn(),
    action: "",
    location: { hash: "1x4v6b8m" },
  };
  props.location = { hash: "1x4v6b8m" };
};

beforeEach(() => {
  setupProps();
});

describe("ScrollSideMenu", () => {
  it("renders properly", () => {
    const component = shallow(<ScrollSideMenu {...props} />);
    expect(component).toMatchSnapshot();
  });
});
