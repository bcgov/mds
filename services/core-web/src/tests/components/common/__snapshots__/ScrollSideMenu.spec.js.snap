// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollSideMenu renders properly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
